### PR TITLE
BUG: catch invalid fixed-width dtype sizes

### LINKED ISSUE #26556: BUG: numpy.string.expandtabs segfault with extreme tab sizes
### Describe the issue:

I randomly noticed that if I slightly modify the existing test `test_expandtabs_raises_overflow` in `test_strings.py`, I get a segmentation fault.

### Reproduce the code example:

```python
import numpy as np
np.strings.expandtabs(np.array("\t", dtype='S'), 2**61)
```

### Error message:

```shell
Segmentation fault
(gdb) bt
#0  __memset_avx2_unaligned_erms () at ../sysdeps/x86_64/multiarch/memset-vec-unaligned-erms.S:328
#1  0x00007f5143c9fab9 in int string_expandtabs_loop<(ENCODING)0>(PyArrayMethod_Context_tag*, char* const*, long const*, long const*, NpyAuxData_tag*) ()
   from /home/msimacek/tmp/cpython-venv/lib64/python3.12/site-packages/numpy/_core/_multiarray_umath.cpython-312-x86_64-linux-gnu.so
#2  0x00007f5143c60bed in try_trivial_single_output_loop (context=context@entry=0x7ffe7410d600, op=op@entry=0x7ffe7410cc00, order=order@entry=NPY_KEEPORDER, errormask=521)
    at ../numpy/_core/src/umath/ufunc_object.c:985
#3  0x00007f5143c6b46e in PyUFunc_GenericFunctionInternal (wheremask=<optimized out>, order=<optimized out>, casting=<optimized out>, op=0x7ffe7410cc00, operation_descrs=0x7ffe7410d000, 
    ufuncimpl=<optimized out>, ufunc=<optimized out>) at ../numpy/_core/src/umath/ufunc_object.c:2257
#4  ufunc_generic_fastcall (ufunc=<optimized out>, args=<optimized out>, len_args=<optimized out>, kwnames=<optimized out>, outer=<optimized out>) at ../numpy/_core/src/umath/ufunc_object.c:4533
#5  0x00007f515220746c in PyObject_Vectorcall () from /lib64/libpython3.12.so.1.0
#6  0x00007f515210f7c7 in _PyEval_EvalFrameDefault.cold () from /lib64/libpython3.12.so.1.0
#7  0x00007f5152278b36 in PyEval_EvalCode () from /lib64/libpython3.12.so.1.0
#8  0x00007f515229edba in run_eval_code_obj () from /lib64/libpython3.12.so.1.0
#9  0x00007f515229947e in run_mod () from /lib64/libpython3.12.so.1.0
#10 0x00007f51521defec in PyRun_InteractiveOneObjectEx () from /lib64/libpython3.12.so.1.0
#11 0x00007f51521df1ea in _PyRun_InteractiveLoopObject () from /lib64/libpython3.12.so.1.0
#12 0x00007f51521515a3 in _PyRun_AnyFileObject.cold () from /lib64/libpython3.12.so.1.0
#13 0x00007f51521df3e2 in PyRun_AnyFileExFlags () from /lib64/libpython3.12.so.1.0
#14 0x00007f51521469c0 in Py_RunMain.cold () from /lib64/libpython3.12.so.1.0
#15 0x00007f515226100c in Py_BytesMain () from /lib64/libpython3.12.so.1.0
#16 0x00007f5151e4614a in __libc_start_call_main (main=main@entry=0x55935bf81160 <main>, argc=argc@entry=1, argv=argv@entry=0x7ffe74111f78) at ../sysdeps/nptl/libc_start_call_main.h:58
#17 0x00007f5151e4620b in __libc_start_main_impl (main=0x55935bf81160 <main>, argc=1, argv=0x7ffe74111f78, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffe74111f68)
    at ../csu/libc-start.c:360
#18 0x000055935bf81095 in _start ()
```

### Python and NumPy Versions:

1.26.4
3.12.3 (main, Apr 17 2024, 00:00:00) [GCC 13.2.1 20240316 (Red Hat 13.2.1-7)]

### Runtime Environment:

_No response_

### Context for the issue:

I'm a developer of GraalPy and I was originally trying to execute numpy's test suite on GraalPy. GraalPy has lower `sys.maxsize`, so it encounters the segfault in the aforementioned test even without modifications. We'd like to be able to run all the tests on GraalPy.

### PR DESCRIPTION
**Pull request by ngoldbaum**
Fixes #26556 

The ultimate cause was the error checking for invalid dtype names like `'S' + str(2**61)` was incorrect. The root cause in the issue is it cast from long to int without doing any kind of overflow detection.

I also noticed another issue while I was debugging: the docs for `strtol` say that it returns 0 if the string isn't representable as an int and we need to check for that case too. It's a little tricky to do that, because we also allow types like `'S0'`, which is why the check is written in a kind of weird way with a gnarly if statement. Very open to alternative ways to write that.

Also noticed that `expandtabs` was assuming `PY_SSIZE_T_MAX` is the largest representable string, but that's wrong, it's INT_MAX because of the conversion form `long` to `int` on the result of `strtol`.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
@seberg @mhvk @lysnikolaou do one of you want to take a look at this?

File: numpy/_core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -1813,15 +1813,30 @@ _convert_from_str(PyObject *obj, int align)
...
+        elsize = (int)result;
+
+
         if (typeend - type == len) {
 
             kind = type[0];
+            // if it's a type + size dtype and the parsed size is zero
+            // and the size of the typestring is longer than 1 character
+            // then strtol must have returned 0 indicating an error
+            if ((elsize == 0) && (typeend > type + 2) &&
```
Comment (mhvk) [inline numpy/_core/src/multiarray/descriptor.c:1831]: Would `S00000` now fail? From the description of `strtol`, it also seems that in your example of `S123hello`, `typeend - type` would not be `len`.

Comment (ngoldbaum) [reply numpy/_core/src/multiarray/descriptor.c:1831]: Looks like it:

```
In [1]: np.dtype('S0000000')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 np.dtype('S0000000')

TypeError: data type 'S0000000' not understood
```

Let me see if I can find a better heuristic that works with that test case...

Comment (ngoldbaum) [reply numpy/_core/src/multiarray/descriptor.c]: Ended up deciding the `typeend - type == len` check is sufficient and getting rid of the `strtol` error checking.

File: numpy/_core/src/multiarray/descriptor.c [hunk 2]
```diff
@@ -1813,12 +1813,15 @@ _convert_from_str(PyObject *obj, int align)
...
+        long result = strtol(type + 1, &typeend, 10);
+        // make sure it doesn't overflow or go negative
+        if (result > INT_MAX || result < 0) {
             goto fail;
         }
 
+        elsize = (int)result;
+
+
         if (typeend - type == len) {
```
Comment (mhvk) [inline numpy/_core/src/multiarray/descriptor.c:1825]: Shouldn't an error be raised if this is not true?

Comment (ngoldbaum) [reply numpy/_core/src/multiarray/descriptor.c:1825]: Right now it falls through to `goto fail` eventually but I can see if anything breaks if add an `else {goto fail;}` block.

**Comment by ngoldbaum** [comment]
Thanks for the review! 

It turns out you do need to do error checking on strtol to fail early, which you can get by checking `errno` after the operation. We still need to make sure we're not passed a string like `"bool"` at this point, but the `typeend - type` check is sufficient to exclude that.

Also added a test that leading zeros in dtype names are parsed correctly. Not particularly because I want to support that but more to avoid breaking it unintentionally in the future like I accidentally did in this PR before Marten's review.

**Comment by ngoldbaum** [comment]
Or not, looks like I was unintentionally using a BSD-ism on my Mac....

**Pull request merged**